crates/client/gateway/src/client/builder.rs

```diff
@@ -48,4 +76,170 @@ impl FeederClient {
                 .expect("Failed to parse Starknet Alpha Sepolia feeder gateway url. This should not fail in prod."),
         )
     }
+
+    pub async fn request_with_pause_handling<F, R>(&self, f: F) -> Result<R, Box<dyn Error + Send + Sync>>
+    where
+        F: FnOnce() -> R,
+    {
+        let pause_until = self.pause_until.clone();
+
+        // Check if a pause is active
+        let pause_duration = {
+            let lock = pause_until.read().await;
+            if let Some(pause_instant) = *lock {
+                let now = Instant::now();
+                if pause_instant > now {
+                    Some(pause_instant - now)
+                } else {
+                    *pause_until.write().await = None;
+                    None
+                }
+            } else {
+                None
+            }
+        };
+
+        // Wait for the pause duration if needed
+        if let Some(duration) = pause_duration {
+            tokio::time::sleep(duration).await;
+        }
+
+        Ok(f())
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct RetryPolicy {
+    max_retries: usize,
+    backoff: Duration,
+    pause_until: Arc<RwLock<Option<Instant>>>,
+}
+
+impl RetryPolicy {
+    pub fn new(max_retries: usize, backoff: Duration, pause_until: Arc<RwLock<Option<Instant>>>) -> Self {
+        RetryPolicy { max_retries, backoff, pause_until }
+    }
+}
+
+impl Policy<Request<String>, Response<Incoming>, Box<dyn Error + Send + Sync>> for RetryPolicy {
+    type Future = Pin<Box<dyn Future<Output = Self> + Send>>;
+
+    fn retry(
+        &self,
+        _: &Request<String>,
+        result: Result<&Response<Incoming>, &Box<dyn Error + Send + Sync>>,
+    ) -> Option<Self::Future> {
+        let pause_until = self.pause_until.clone();
+
+        match result {
+            Ok(response) => {
+                // TOO_MANY_REQUESTS
+                if response.status().as_u16() == 429 {
```

Comment:
Consider using `http::StatusCode::TOO_MANY_REQUESTS` instead.

```suggestion
                if response.status() == http::StatusCode::TOO_MANY_REQUESTS {
```